edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/statemachine/EndpointExecutorFsm.cs

```diff
@@ -286,6 +290,15 @@ static async Task EnterSendingAsync(EndpointExecutorFsm thisPtr)
             ICollection<IMessage> messages = EmptyMessages;
             Stopwatch stopwatch = Stopwatch.StartNew();
             TimeSpan endpointTimeout = TimeSpan.FromMilliseconds(thisPtr.config.Timeout.TotalMilliseconds * thisPtr.Endpoint.FanOutFactor);
+
+            if (thisPtr.EndpointExecutorToken.IsCancellationRequested)
+            {
+                Events.SendCanceled(thisPtr);
+                next = Commands.Checkpoint(SinkResult<IMessage>.Empty);
```

Comment:
is it ok to set next state to empty checkpoint? and do I still need to call RunInternalAsync?

Reply:
Something like Commands.Die is probably more appropriate, since at this point the endpoint is gone and these messages should just get dropped anyways

Either way I think you still do need to call RunInternalAsync, or the original command that the caller is waiting on won't get completed.

---
In reply to: [412575655](https://github.com/Azure/iotedge/pull/2865#discussion_r412575655) [](ancestors = 412575655)

Reply:
Yes, we probably need to discard all messages